sdk/src/Services/S3/Custom/Transfer/Internal/PartBufferManager.cs

```diff
@@ -0,0 +1,282 @@
+/*******************************************************************************
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use
+ *  this file except in compliance with the License. A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file.
+ *  This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ *  CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *  specific language governing permissions and limitations under the License.
+ * *****************************************************************************
+ *    __  _    _  ___
+ *   (  )( \/\/ )/ __)
+ *   /__\ \    / \__ \
+ *  (_)(_) \/\/  (___/
+ *
+ *  AWS SDK for .NET
+ *  API Version: 2006-03-01
+ *
+ */
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Amazon.S3.Transfer.Internal
+{
+    /// <summary>
+    /// Manages part buffers with ArrayPool lifecycle and concurrency control.
+    /// Uses IPartDataSource for uniform handling of streaming and buffered data.
+    /// </summary>
+    internal class PartBufferManager : IPartBufferManager
+    {
+        private readonly ConcurrentDictionary<int, IPartDataSource> _partDataSources;
+        private readonly SemaphoreSlim _bufferSpaceAvailable;
```

Comment:
Using 3 different lock primitives makes this pretty confusing. I'm not saying they aren't needed but this type of complexity should be documented here. For example in recent revamp of the locking in refreshing credentials I gave an explanation in code how the locking was being used to help future readers. https://github.com/aws/aws-sdk-net/blob/main/sdk/src/Core/Amazon.Runtime/Credentials/RefreshingAWSCredentials.cs#L171

Reply:
i will add comments here